### ISSUE TITLE
RDotNet raises dangerous AccessViolationException

### ISSUE THREAD

**Issue by hrogithub** [rdotnet/rdotnet#133]
Any help available,

we have this trouble since many versions of RDotNet and R.
Current version is
Windows Server 2016 64 bit RDotNet-1.8.2 R-4.0.2 .Net-4.7.2

Under special data and call sequence circumsances it is reproduceable, that RDotNet calls crashes with System.AccessViolationException.

We build a complex vector with lists and scalar values as SymbolicExpressions as an parameter for our R-Function call.
Under rare, but if, than reproduceable conditions, this build raise the access violation!
The creation of a symbolic expression will fail.
If you change any sequence for creation, or if any data content is different, all works fine...

So it is not possible to extract this case as an simple reproduceable example.

Reason:
It seems to dealing with garbage collection of DotNet runtime and R-Runtime.
R-Runtime will compact/move memory and during this it will possibly corrupt .Net pointer objects?

One solution we found is to call RDotNet.REngine.ForceGarbageCollection() after each creation of any symbnolic expression,
but this will seriously impact the performance 
and as you can see, it is under different conditions the wrong decission and without force GC all works fine.

What is  the safe way to create complex nested symbolic expressions to avoid this effects?
Best would be to stop both GC processes during Interaction between RDotNet and R-Runtime, but how?

Exception:
ServiceAgentCommand  Machine Learning Service raised an exception 'Domain First Chance Exception'. Message: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.Critical Exception has been raised: System.AccessViolationException
System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at RDotNet.REngine.ForceGarbageCollection()

Tnx

Roman

**Comment by lrasmus** [comment]
Thanks for the information - just to clarify, there isn't a reproducible example that you could share?  Or maybe some example code in which you've seen this occur (even if it sometimes works)?

**Comment by hrogithub** [comment]
Hi Irasmus,
as I mentioned, I can not share our full app to reproduce the error and I can not isolate the error in a small sample.

System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at RDotNet.Vector`1..ctor(REngine engine, SymbolicExpressionType type, IEnumerable`1 vector)

2020-10-21 12:02:48,768 12 ERROR   ServiceAgentCommand  Machine Learning Service raised an exception 'Domain Unhandled Exception'. Message: SystemException occurred. Failed to execute script.
RDotNet.EvaluationException: SystemException occurred. Failed to execute script. ---> System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at RDotNet.Vector`1..ctor(REngine engine, SymbolicExpressionType type, IEnumerable`1 vector)
   at RDotNet.NumericVector..ctor(REngine engine, IEnumerable`1 vector)
   at RDotNet.REngineExtension.CreateNumericVector(REngine engine, IEnumerable`1 vector)


The exception will be raced only after a special sequence of REngine calls during creation of parameter data.
It strongly depends on the call order and the call data that force to raise this exception.
Same call sequence with different data works fine.
So 99.99 % works, but some rare and different cases reproduceable not.

Can I setup more trace options to the REngine to better catch this raise condition?
What are the dont's during creation of nested SymbolicExpressions?
I have redirected the R-console output to our logging system, but could not find something about this. have you any clue to look for in that huge output?

Roman

**Comment by joaofabiobh** [comment]
Hi, I'm also having this problem.

I have a service using a Singleton to call one instance of R Engine from a main loop, therefore no multi-thread issues here.
I can't reproduce the problem, but I randomly get this exception.

In my case, I don't care for the SymbolicExpressions, since my script only outputs some charts to SVG files. 
I tried to dispose the SymbolicExpressions, doing:

            SymbolicExpression output = engine.Evaluate(expression);
            output.Dispose();

Still, no luck. Is there a way to evaluate expressions in void methods to those who won't use SymbolicExpressions?

As of now, I'll try to capture the exceptions using this method: https://stackoverflow.com/questions/3469368/how-to-handle-accessviolationexception

Any new information I'll share here.

JF

**Comment by hrogithub** [comment]
Next update:
First the good news,
I found a good workaround to avoiding this problem.
After reading of memory.c and startup.c from R core implementation 
I found out, that the feature to deactivate GC and explain this as an function is planned, but not available.
So I tried to find out, how to minimize the chance of a running GC.

All I have done was increasing the minimum values for VSIZE and NSIZE. (MinCellSize and MinMemorySize)
Couse it does not work with StartupParameter, I used the environment variables
R_VSIZE and R_NSIZE and in addition played with R_GC_MEM_GOW 3 (more aggressive)
But be aware: maximum of R_NSIZE is 50000000 !!! otherwise the value will be ignorred.

Bad news is, I could only suppose, the reason is that: R-Runtime decides to move the SymbolicExpression-Pointer before it is locked by the RDotNET with Preserve() call, 
(this call should be moved before all cast stuff to call it asap in the SymbolicExpression constructor)

Roman
